src/MySqlConnector/MySqlClient/ConnectionPool.cs

```diff
@@ -14,6 +14,13 @@ internal sealed class ConnectionPool
 		{
 			cancellationToken.ThrowIfCancellationRequested();
 
+			// if all sessions are used, see if any have been leaked and can be recovered
+			if (m_sessionSemaphore.CurrentCount == 0)
+			{
+				GC.Collect();
+				RecoverLeakedSessions();
```

Comment:
> perform the full GC and scan only if a timeout exception is thrown so we try to auto-recover only in a completely-leaked pool situation

This approach doesn't really work because the cancellation token is now permanently cancelled (because the connect timeout has elapsed) so all further operations with it abort immediately.